iceoryx_examples/icedelivery/iox_subscriber_untyped_modern.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/modern_api/subscriber.hpp"
+#include "iceoryx_posh/popo/guard_condition.hpp"
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "topic_data.hpp"
+
+#include <chrono>
+#include <csignal>
+#include <iostream>
+
+bool killswitch = false;
+iox::popo::GuardCondition shutdownGuard;
+
+static void sigHandler(int f_sig [[gnu::unused]])
+{
+    // caught SIGINT, now exit gracefully
+    killswitch = true;
+    shutdownGuard.trigger(); // unblock waitsets
+}
+
+void subscriberHandler(iox::popo::WaitSet& waitSet)
+{
+    // run until interrupted
+    while(!killswitch)
+    {
+        auto triggeredConditions = waitSet.wait();
+        for(auto& condition : triggeredConditions)
+        {
+            auto untypedSubscriber = dynamic_cast<iox::popo::UntypedSubscriber*>(condition);
+            if(untypedSubscriber)
+            {
+                untypedSubscriber->take().and_then([](iox::cxx::optional<iox::popo::Sample<const void>>& maybeSample){
```

Comment:
We need to simplify this. I would recommend a signature like:
```cpp
ReceivedSample<const T> take() noexcept;
```

The `ReceivedSample` should then get an API like (it is just a draft to get an idea):
```cpp
template<typename T>
class ReceivedSample {
    void and_then(std::function<void(T&)> ); // is only called when the expected contains a value and the optional contains a value
    // is called when the expected contains an error or the optional does not have a value. 
    // The user has to determine the case and therefore the ReceivedSample is provided here as reference
    void or_else(std::function<void(ReceivedSample &)>); 
    bool hasError(); // returns true if the expected contains an error
    bool hasValue();; // returns true if the expected contains a value as well as the optional
};
```

@ithier @elBoberido @budrus what do you think?

Reply:
This is an idea we should discuss. This could hide the expected and optional behind the scenes, so having a simpler API but not loosing the benefits of expected and optional which are both valid choices for the take

Reply:
The `getValue()` is missing. and this would return a T& ?

Reply:
Added #340 to explore this further.